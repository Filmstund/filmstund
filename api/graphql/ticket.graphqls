extend type Showing {
    myTickets: [Ticket!]!
    ticketRange: TicketRange
}

extend type Mutation {
    processTicketUrls(showingID: UUID!, ticketUrls: [String!]): Showing! @auth(requires: SHOWING_ADMIN)
}

# A ticket assigned to a user that allows access to Filmstaden
type Ticket {
    id: String!
    showingID: UUID!
    assignedToUser: UUID!
    profileID: String
    barcode: String!
    customerType: String!
    customerTypeDefinition: String!
    cinema: String!
    cinemaCity: String
    screen: String!
    seat: Seat!
    date: LocalDate!,
    time: String!,
    movieName: String!
    # 15 år, 11 år etc.
    movieRating: String!
    # "textad", "en" etc
    attributes: [String!]!
}

type Seat {
    row: Int!
    number: Int!
}

type SeatRange {
    row: Int!
    # All numbers for this particular row
    numbers: [Int!]!
}

# In what range is the seating assigned for each row
type TicketRange {
    rows: [Int!]!
    seatings: [SeatRange!]!
    # The total number of tickets for this showing
    totalCount: Int!
}
