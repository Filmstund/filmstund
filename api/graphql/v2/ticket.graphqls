extend type Showing {
    myTickets: [Ticket!]!
    ticketRange: TicketRange
}

# A ticket assigned to a user that allows access to Filmstaden
type Ticket {
    id: String!
    showingID: UUID!
    assignedToUser: UserID!
    profileID: String
    barcode: String!
    customerType: String!
    customerTypeDefinition: String!
    cinema: String!
    cinemaCity: String
    screen: String!
    seat: Seat!
    date: LocalDate!,
    time: String!,
    movieName: String!
    # 15 år, 11 år etc.
    movieRating: String!
    # "textad", "en" etc
    attributes: [String!]!
}

type Seat {
    row: Int!
    number: Int!
}

type SeatRange {
    row: Int!
    # All numbers for this particular row
    numbers: [Int!]!
}

# In what range is the seating assigned for each row
type TicketRange {
    rows: [Int!]!
    seatings: [SeatRange!]!
    # The total number of tickets for this showing
    totalCount: Int!
}
