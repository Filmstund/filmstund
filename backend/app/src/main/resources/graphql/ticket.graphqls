# A ticket assigned to a user that allowes access to Filmstaden
type Ticket {
    id: String!
    #showingId: UUID! FIXME use dto where this exists instead
    assignedToUser: User!
    profileId: String
    barcode: String!
    customerType: String!
    customerTypeDefinition: String!
    cinema: String!
    cinemaCity: String
    screen: String!
    seatNumber: Int!,
    seatRow: Int!,
    date: LocalDate!,
    time: String!,
    movieName: String!
    # 15 år, 11 år etc.
    movieRating: String!
    # "textad", "en" etc
    attributes: [String!]
}

# The row and seat number for a ticket
type Seat {
    row: Int!
    number: Int!
}

type SeatRange {
    row: Int!
    # All numbers for this particular row
    numbers: [Int!]
}

# In what range is the seating assigned for each row
type TicketRange {
    rows: [Int!]
    seatings: [SeatRange!]
    # The total number of tickets for this showing
    totalCount: Int!
}

extend type Showing {
    myTickets: [Ticket!]
    ticketRange: TicketRange
}