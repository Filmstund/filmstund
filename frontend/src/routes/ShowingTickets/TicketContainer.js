import React, { Component } from "react";
import { getJson, jsonRequest } from "../../lib/fetch";
import { withBaseURL } from "../../lib/withBaseURL";
import Loader from "../../ProjectorLoader";
import TicketURLInput from "../../TicketURLInput";
import Field from "../../Field";
import MainButton from "../../MainButton";
import Ticket from "./Ticket";
import RangeDisplay from "./RangeDisplay";

export default class TicketContainer extends Component {
  state = {
    tickets: null,
    error: null,
    range: null,
    cinemaTicketUrls: []
  };

  componentWillMount() {
    getJson(`/tickets/${this.props.showingId}/seating-range`).then(range => {
      this.setState({ range });
    });
    getJson(`/tickets/${this.props.showingId}`).then(
      tickets => {
        this.setState({ tickets });
      },
      err => {
        this.setState({
          error: "Ingen biljett registrerad för köpet"
        });
      }
    );
  }

  handleSubmitCinemaTicketUrls = () => {
    const { cinemaTicketUrls } = this.state;
    const { showing } = this.props;

    const nonEmptyUrls = cinemaTicketUrls.filter(
      line => line.trim().length !== 0
    );

    jsonRequest(withBaseURL(`/tickets/${showing.id}`), nonEmptyUrls).then(
      tickets => {
        this.setState({
          tickets,
          cinemaTicketUrls: []
        });
      }
    );
  };

  setCinemaTicketUrls = cinemaTicketUrls => {
    this.setState({
      cinemaTicketUrls
    });
  };

  renderAdminFields = () => {
    const { cinemaTicketUrls } = this.state;

    return (
      <Field text="Lägg till SF-biljettlänkar">
        <TicketURLInput
          cinemaTicketUrls={cinemaTicketUrls}
          onChange={this.setCinemaTicketUrls}
        />
        <MainButton onClick={this.handleSubmitCinemaTicketUrls}>
          Skicka
        </MainButton>
      </Field>
    );
  };

  handleGoBackToShowing = () => {
    this.props.history.goBack();
  };

  render() {
    const { range, tickets, error } = this.state;
    const { me, showing } = this.props;

    if (error) {
      return (
        <div>
          <MainButton onClick={this.handleGoBackToShowing}>
            Tillbaka till visning
          </MainButton>
          <div>{error}</div>
        </div>
      );
    } else if (!tickets) {
      return <Loader />;
    } else {
      return (
        <div>
          <MainButton onClick={this.handleGoBackToShowing}>
            Tillbaka till visning
          </MainButton>
          {range && <RangeDisplay range={range} />}
          {tickets.map(ticket => <Ticket key={ticket.id} {...ticket} />)}
          {showing.admin === me.id && this.renderAdminFields()}
        </div>
      );
    }
  }
}
