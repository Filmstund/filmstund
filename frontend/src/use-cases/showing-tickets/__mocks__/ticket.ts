import {
  Ticket_filmstadenSeatMap,
  Ticket_myTickets
} from "../../../apollo/mutations/__generated__/Ticket";

export const createMockSeat = (
  row: number,
  number: number,
  x: number,
  y: number
): Ticket_filmstadenSeatMap => ({
  __typename: "FilmstadenSeatMapDTO",
  seatType: "TEST",
  row,
  number,
  coordinates: {
    __typename: "FilmstadenSeatCoordinates",
    x,
    y
  },
  dimensions: {
    __typename: "FilmstadenSeatDimensions",
    width: 10,
    height: 10
  }
});

export const createMockTicket = (
  id: string,
  row: number,
  number: number
): Ticket_myTickets => ({
  __typename: "TicketDTO",
  id,
  barcode: "this-is-barcode",
  customerType: "mock-customer-type",
  customerTypeDefinition: "mock-customer-type-definition",
  cinema: "mock-cinema",
  screen: "mock-screen",
  profileId: "mock-profile-id",
  seat: {
    __typename: "Seat",
    row,
    number
  },
  date: "2018-10-10",
  time: "20:00",
  movieName: "mock-movie-name",
  movieRating: "R",
  attributes: ["3D"]
});
