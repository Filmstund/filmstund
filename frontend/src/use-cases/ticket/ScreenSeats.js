import React from "react";
import styled from "@emotion/styled";
import { minBy, maxBy, flatMap } from "lodash-es";

const Seat = styled.div(({ selected }) => ({
  backgroundColor: selected ? "#d0021b" : "#9b9b9b",
  position: "absolute"
}));

const Screen = styled.div`
  position: relative;
  height: ${props => props.height}px;
  overflow: scroll;
`;

export const ScreenSeats = ({ ticketRange, seatMap }) => {
  if (seatMap.length === 0) {
    return null;
  }

  const heightRatio = 351 / 464;

  const maxWidth = Math.min(window.innerWidth - 80, 464);
  const maxHeight = maxWidth * heightRatio;

  const width = 10;
  const height = 10;

  const minY = minBy(seatMap, s => s.coordinates.y).coordinates.y;

  const minX = minBy(seatMap, s => s.coordinates.x).coordinates.x;

  const maxX = maxBy(seatMap, s => s.coordinates.x).coordinates.x + width;
  const maxY = maxBy(seatMap, s => s.coordinates.y).coordinates.y + height;

  const scaleX = maxWidth / (maxX - minX);
  const scaleY = maxHeight / (maxY - minY);

  const numbers = flatMap(ticketRange.seatings, seating => seating.numbers);

  return (
    <Screen height={maxHeight}>
      {seatMap.map(data => (
        <Seat
          style={{
            top: parseInt((data.coordinates.y - minY) * scaleY, 10),
            left: parseInt((data.coordinates.x - minX) * scaleX, 10),
            height: parseInt(height * scaleY, 10),
            width: parseInt(width * scaleX, 10)
          }}
          key={data.number}
          selected={numbers.includes(data.number)}
        />
      ))}
    </Screen>
  );
};
