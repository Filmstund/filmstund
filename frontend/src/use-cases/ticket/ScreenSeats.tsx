import styled from "@emotion/styled";
import { flatMap, groupBy, map } from "lodash-es";
import React from "react";
import {
  TicketQuery_showing_sfSeatMap,
  TicketQuery_showing_ticketRange
} from "../showing-tickets/__generated__/TicketQuery";

const Seat = styled.div<{ selected: boolean }>(({ selected }) => ({
  backgroundColor: selected ? "#d0021b" : "#9b9b9b",
  position: "absolute",
  borderRadius: 2
}));

const Screen = styled.div<{ height: number; width: number }>`
  margin: 20px 0;
  position: relative;
  height: ${props => props.height}px;
  width: ${props => props.width}px;
`;

interface Props {
  ticketRange: TicketQuery_showing_ticketRange;
  seatMap: TicketQuery_showing_sfSeatMap[];
}

export const ScreenSeats: React.FC<Props> = ({ ticketRange, seatMap }) => {
  if (seatMap.length === 0) {
    return null;
  }

  const rows = groupBy(seatMap, s => s.row);
  const widestRowLength = Math.max(...map(rows, row => row.length));

  const width = Math.min(window.innerWidth - 40, window.innerHeight, 650);

  const optimalSeatWidth = 17;
  const seatWidth = Math.min(optimalSeatWidth, width / (widestRowLength * 1.2));
  const seatHeight = Math.floor(13 * seatWidth / optimalSeatWidth);

  const minY = Math.min(...seatMap.map(s => s.coordinates.y));
  const maxY = Math.max(...seatMap.map(s => s.coordinates.y));
  const minX = Math.min(...seatMap.map(s => s.coordinates.x));
  const maxX = Math.max(...seatMap.map(s => s.coordinates.x));

  const scale = width / (maxX - minX);

  const numbers = flatMap(
    ticketRange.seatings,
    seating => seating.numbers || []
  );

  return (
    <Screen
      height={(maxY - minY) * scale + seatHeight}
      width={(maxX - minX) * scale + seatWidth}
    >
      {seatMap.map(data => (
        <Seat
          style={{
            left: Math.floor((data.coordinates.x - minX) * scale),
            top: Math.floor((data.coordinates.y - minY) * scale),
            height: seatHeight,
            width: seatWidth
          }}
          key={data.number}
          selected={numbers.includes(data.number)}
        />
      ))}
    </Screen>
  );
};
