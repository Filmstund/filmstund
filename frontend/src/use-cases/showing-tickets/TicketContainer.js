import React, { Component } from "react";
import Loader from "../../use-cases/common/utils/ProjectorLoader";
import TicketURLInput from "../../use-cases/common/ui/TicketURLInput";
import MainButton from "../../use-cases/common/ui/MainButton";
import Ticket from "./Ticket";
import SeatRange from "./SeatRange";
import StatusMessageBox from "../../use-cases/common/utils/StatusMessageBox";
import { PageWidthWrapper } from "../../use-cases/common/ui/PageWidthWrapper";
import { ScreenSeats } from "../../use-cases/ticket/ScreenSeats";
import { SmallHeader } from "../../use-cases/common/ui/Header";
import { navigateToShowing } from "../common/navigators/index";

import { FieldWithoutMaxWidth } from "../common/ui/Field";

export default class TicketContainer extends Component {
  state = {
    errors: null,
    success: false,
    cinemaTicketUrls: []
  };

  handleSubmitCinemaTicketUrls = () => {
    const { cinemaTicketUrls } = this.state;
    const {
      data: { showing }
    } = this.props;

    const nonEmptyUrls = cinemaTicketUrls.filter(
      line => line.trim().length !== 0
    );

    this.props
      .addTickets(showing.id, nonEmptyUrls)
      .then(() => {
        this.setState({ success: true, errors: null, cinemaTicketUrls: [] });
      })
      .catch(errors => {
        this.setState({ success: false, errors });
      });
  };

  setCinemaTicketUrls = cinemaTicketUrls => {
    this.setState({
      cinemaTicketUrls
    });
  };

  renderAdminFields = () => {
    const { cinemaTicketUrls } = this.state;

    return (
      <FieldWithoutMaxWidth text="Lägg till SF-bokningslänkar">
        <TicketURLInput
          cinemaTicketUrls={cinemaTicketUrls}
          onChange={this.setCinemaTicketUrls}
        />
        <MainButton onClick={this.handleSubmitCinemaTicketUrls}>
          Skicka
        </MainButton>
      </FieldWithoutMaxWidth>
    );
  };

  handleGoBackToShowing = () => {
    const {
      history,
      data: { showing }
    } = this.props;
    navigateToShowing(history, showing);
  };

  render() {
    const { errors, success } = this.state;
    const {
      data: { me, showing, loading }
    } = this.props;

    const { myTickets, ticketRange, sfSeatMap } = showing;

    if (loading) {
      return <Loader />;
    } else {
      return (
        <PageWidthWrapper>
          <MainButton onClick={this.handleGoBackToShowing}>
            Tillbaka till visning
          </MainButton>
          <SmallHeader>Våra platser:</SmallHeader>
          <SeatRange ticketRange={ticketRange} />
          <ScreenSeats ticketRange={ticketRange} seatMap={sfSeatMap} />
          {myTickets.map(ticket => (
            <Ticket key={ticket.id} {...ticket} />
          ))}
          <StatusMessageBox
            success={success}
            errors={errors}
            successMessage="Uppdaterades!"
          />
          {showing.admin.id === me.id && this.renderAdminFields()}
        </PageWidthWrapper>
      );
    }
  }
}
