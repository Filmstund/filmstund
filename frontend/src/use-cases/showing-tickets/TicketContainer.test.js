import React from "react";
import { render, cleanup } from "@testing-library/react";

import {TicketContainer} from "./TicketContainer";
import { createMockTicket, createMockSeat } from "./__mocks__/ticket";

const mockUserData = {
  id: "this-is-mock-user-id"
};

const minimalMockData = {
  me: mockUserData,
  showing: {
    id: "this-is-mock-showing-id",
    admin: mockUserData,
    ticketRange: {
      rows: [],
      seatings: []
    },
    sfSeatMap: [],
    myTickets: [createMockTicket(1, 2, 3)]
  }
};

const mockData = {
  me: mockUserData,
  showing: {
    id: "this-is-mock-showing-id",
    admin: mockUserData,
    ticketRange: {
      rows: [2, 3],
      seatings: [
        {
          row: 2,
          numbers: [3, 4]
        },
        {
          row: 3,
          numbers: [5, 6]
        }
      ]
    },
    sfSeatMap: [
      createMockSeat(2, 3, 5, 0),
      createMockSeat(2, 4, 10, 0),
      createMockSeat(3, 5, 5, 5),
      createMockSeat(3, 6, 10, 5)
    ],
    myTickets: [
      createMockTicket(1, 2, 3),
      createMockTicket(2, 2, 4),
      createMockTicket(3, 3, 5),
      createMockTicket(4, 3, 6)
    ]
  }
};

afterEach(cleanup);

describe("<TicketContainer />", () => {
  it("renders without crashing", () => {
    const { container } = render(<TicketContainer data={minimalMockData} />);
    expect(container).toMatchSnapshot();
  });

  it("renders the correct amount of tickets", () => {
    const { queryAllByText } = render(<TicketContainer data={mockData} />);
    expect(queryAllByText("mock-cinema")).toHaveLength(4);
  });
});
