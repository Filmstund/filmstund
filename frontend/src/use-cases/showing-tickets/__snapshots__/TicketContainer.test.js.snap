// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicketContainer /> renders without crashing 1`] = `
<div>
  <div
    class="css-pac1zl"
  >
    <button
      class="css-gcz4q9"
    >
      Tillbaka till visning
    </button>
    <h3
      class="css-1llnadu"
    >
      Våra platser:
    </h3>
    <div
      style="display: flex; justify-content: center;"
    />
    <div
      class="css-18sytmz"
    >
      <div
        class="css-kxhk2b-CompanyHeader"
      >
        <img
          alt="FS logo"
          class="css-x7d99d-CompanyHeader"
          src="fslogo.png"
        />
        <div
          class="css-deizzc"
        >
          <div
            class="css-15wq1tr-CompanyHeader"
          >
            mock-cinema
          </div>
          <div
            class="css-e2s1aj"
          >
            Filmstaden
          </div>
        </div>
      </div>
      <div
        class="css-1d99se3-TicketHeader"
      >
        <div
          class="css-1k4w5qa-TicketHeader"
        >
          mock-movie-name
        </div>
        <div
          class="css-f24dyb"
        >
          <div
            class="css-o2j9ze"
          >
            <span
              class="css-ya7yf4"
            >
              3D
            </span>
          </div>
          R
        </div>
      </div>
      <div
        class="css-1dvy66q-TicketDateTime"
      >
        <div
          class="css-18notvi-TicketValueWithLabel"
        >
          <div
            class="css-e2s1aj"
          >
            Datum
          </div>
          <div
            class="css-lp25y4"
          >
            2018-10-10
          </div>
        </div>
        <div
          class="css-18notvi-TicketValueWithLabel"
        >
          <div
            class="css-e2s1aj"
          >
            Tid
          </div>
          <div
            class="css-lp25y4"
          >
            20:00
          </div>
        </div>
      </div>
      <div
        class="css-okouzr-TicketPlacement"
      >
        <div
          class="css-18notvi-TicketValueWithLabel"
        >
          <div
            class="css-e2s1aj"
          >
            Salong
          </div>
          <div
            class="css-lp25y4"
          >
            mock-screen
          </div>
        </div>
        <div
          class="css-18notvi-TicketValueWithLabel"
        >
          <div
            class="css-e2s1aj"
          >
            Rad
          </div>
          <div
            class="css-lp25y4"
          >
            2
          </div>
        </div>
        <div
          class="css-18notvi-TicketValueWithLabel"
        >
          <div
            class="css-e2s1aj"
          >
            Stolsnr
          </div>
          <div
            class="css-lp25y4"
          >
            3
          </div>
        </div>
      </div>
      <div
        class="css-fpbkz6-TicketCustomerType"
      >
        mock-customer-type
      </div>
      <div
        class="css-1vldiem-TicketCode"
      >
        <div
          class="css-11v75ym-TicketCode"
        >
          <img
            alt="1"
            class="css-6n2ytg"
            src="this-is-barcode"
          />
          <div>
            1
             
            mock-profile-id
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-1dhwn8t"
    >
      <label
        class="css-1vtr8wv"
      >
        Lägg till Filmstaden-bokningslänkar
      </label>
      <br />
      <textarea
        class="css-1tvymxn"
        rows="5"
      />
      <button
        class="css-gcz4q9"
      >
        Skicka
      </button>
    </div>
  </div>
</div>
`;
